Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp

```diff
@@ -75,21 +75,26 @@ void RemoteQueue::onSubmittedWorkDone(CompletionHandler<void()>&& callback)
 void RemoteQueue::writeBuffer(
     WebGPUIdentifier buffer,
     PAL::WebGPU::Size64 bufferOffset,
-    Vector<uint8_t>&& data)
+    Vector<uint8_t>&& data,
+    CompletionHandler<void()>&& callback)
```

Comment:
I don't think this is right. Writing a buffer should just queue up the write operation. It doesn't need a callback, and it doesn't need to block the web process.

If the buffer isn't in use, and the buffer is shared, writeBuffer can just be a memcpy. If the buffer is in use, or if it isn't shared, writeBuffer should copy the data into a temporary buffer, and enqueue a blit command to copy the data from the temporary buffer to its final destination.

IIRC I already implemented this logic :/

Reply:
Oh good point I misunderstood the CTS test.

Indeed I think this blocking should be removed. Instead I think we have a bug in Buffer::mapAsync.

Reply:
Specifically `Buffer::getMappedRange` in WebGPU.framework returns `return static_cast<char*>(m_buffer.contents) + offset;` but if there is a queued up blit command on that buffer then that is wrong I think. It will return the buffer's current contents as opposed to the contents after the blit command finishes

Reply:
getMappedRange shouldn't block; instead, mapAsync should be the one blocking until all commands referencing the buffer are finished

Reply:
mapAsync is also using `onSubmittedWorkDone()` 

My current theory is the bug is in `unmap`, but I am investigating.